storage: deflake TestNodeHeartbeatCallback

tbg opened the pull request:
The node could become live by observing the Gossip update (triggered by its
heartbeat write) before getting to the point in `heartbeatInternal` at which
it would invoke the heartbeat callback. As a result, one location in this
test needs to be prepared to wait for a split second.

Flaky in the hundreds before, now not flaky with 2k+ iterations.

Fixes #19362.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/19386)
<!-- Reviewable:end -->

nvanbenschoten commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/storage/node_liveness_test.go, line 230 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19386#-KwqHUigXwlljRwhwBwl:-KwqHUihCpWRf_PQ7w0w:b-rsdy6n) ([raw file](https://github.com/cockroachdb/cockroach/blob/24d30541aea1d6ec80683a42c80e5f9f2f748746/pkg/storage/node_liveness_test.go#L230)):*
> ```Go
> 	}
> 
> 	// We have to retry here since the node could have become live via its own
> ```

The effect of the node becoming live via its own Gossip update is that `mtc.Start` returns, right? So the race is between `mtc.Start` returning so we can call `verifyUptimes` and the heartbeat callback firing?

If so, can we change `mtc.addStore` to only return when the callback has been called? We have direct access to it since we pass the callback into `StartHeartbeat`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19386)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/storage/node_liveness_test.go, line 230 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19386#-KwqHUigXwlljRwhwBwl:-KwqQJ8hau9--YKeI_bC:b-u6is21) ([raw file](https://github.com/cockroachdb/cockroach/blob/24d30541aea1d6ec80683a42c80e5f9f2f748746/pkg/storage/node_liveness_test.go#L230)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

The effect of the node becoming live via its own Gossip update is that `mtc.Start` returns, right? So the race is between `mtc.Start` returning so we can call `verifyUptimes` and the heartbeat callback firing?

If so, can we change `mtc.addStore` to only return when the callback has been called? We have direct access to it since we pass the callback into `StartHeartbeat`.
</blockquote></details>

Not sure I follow -- the nodes become live because they heartbeat their record in the background (as a running server does). The detail that the callback isn't atomic with the heartbeat is inconsequential outside of this test, and it seems preferable to document it in the test rather than hiding the fact somewhere else. Perhaps I'm missing your point?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19386)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/storage/node_liveness_test.go, line 230 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19386#-KwqHUigXwlljRwhwBwl:-KwqVE7qOoa-aQrcWBJN:b-i104y8) ([raw file](https://github.com/cockroachdb/cockroach/blob/24d30541aea1d6ec80683a42c80e5f9f2f748746/pkg/storage/node_liveness_test.go#L230)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Not sure I follow -- the nodes become live because they heartbeat their record in the background (as a running server does). The detail that the callback isn't atomic with the heartbeat is inconsequential outside of this test, and it seems preferable to document it in the test rather than hiding the fact somewhere else. Perhaps I'm missing your point?
</blockquote></details>

I guess what I'm getting at is that in `mtc.addStore` we currently "Wait until we see the first heartbeat" by polling `m.nodeLivenesses[idx].IsLive(nodeID)`. Once we see this, we begin to race with the callback, which causes this flake. We could instead wait until we see the first call to the heartbeat callback (which we set there) and avoid this race. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19386)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: deflake TestNodeHeartbeatCallback

The node could become live by observing the Gossip update (triggered by its
heartbeat write) before getting to the point in `heartbeatInternal` at which
it would invoke the heartbeat callback. As a result, one location in this
test needs to be prepared to wait for a split second.

Flaky in the hundreds before, now not flaky with 2k+ iterations.

Fixes #19362.

tschottdorf commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/storage/node_liveness_test.go, line 230 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19386#-KwqHUigXwlljRwhwBwl:-KwsoxzZBpi7-izSqGqf:b-mp42hu) ([raw file](https://github.com/cockroachdb/cockroach/blob/24d30541aea1d6ec80683a42c80e5f9f2f748746/pkg/storage/node_liveness_test.go#L230)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I guess what I'm getting at is that in `mtc.addStore` we currently "Wait until we see the first heartbeat" by polling `m.nodeLivenesses[idx].IsLive(nodeID)`. Once we see this, we begin to race with the callback, which causes this flake. We could instead wait until we see the first call to the heartbeat callback (which we set there) and avoid this race. 
</blockquote></details>

Ah, I see. Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19386)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
@nvanbenschoten I'll merge this because it's an annoying flake, but please double-check that I'm doing what you had in mind.

nvanbenschoten commented:
LGTM, yeah that's exactly what I meant, thanks!
